rllib/env/multi_agent_episode.py

```diff
@@ -755,6 +753,108 @@ def finalize(
+                # Add the agent's starting timestep.
+                self.agent_t_started[agent_id] = other.agent_t_started[agent_id]
+
+            # If the agent was done in the first episode chunk, continue.
+            elif sa_episode.is_done:
+                continue
+
+            # If the agent has data in both chunks, simply concatenate on the
+            # single-agent level.
+            elif agent_id in other.agent_episodes:
+                # If `self` has hanging agent values -> Add these to `other`'s agent
+                # SingleAgentEpisode (as a new timestep) and only then concatenate.
+                # Otherwise, the concatentaion would fail b/c of missing data.
+                if agent_id in self._agent_buffered_actions:
+                    assert agent_id in self._agent_buffered_extra_model_outputs
+                    sa_episode.add_env_step(
+                        observation=other.agent_episodes[agent_id].get_observations(0),
+                        infos=other.agent_episodes[agent_id].get_infos(0),
+                        action=self._agent_buffered_actions[agent_id],
+                        reward=self._agent_buffered_rewards[agent_id],
+                        extra_model_outputs=(
+                            self._agent_buffered_extra_model_outputs[agent_id]
+                        ),
+                    )
+
+                sa_episode.concat_episode(other.agent_episodes[agent_id])
+                # Concatenate the env- to agent-timestep mappings.
+                last_agent_step = next(
+                    item
+                    for item in reversed(self.env_t_to_agent_t[agent_id])
+                    if item != self.SKIP_ENV_TS_TAG
+                )
+                first_agent_step = next(
+                    item
+                    for item in other.env_t_to_agent_t[agent_id]
+                    if item != other.SKIP_ENV_TS_TAG
+                )
+                self.env_t_to_agent_t[agent_id] += (
+                    other.env_t_to_agent_t[agent_id]
+                    + (last_agent_step - first_agent_step)
+                )[1:]
+
+            # Otherwise, the agent is only in `self` and not done. All data is stored
+            # already -> skip
+            # else: pass
+
+        # Update all timestep counters.
+        self.env_t = other.env_t
+        # Check, if the episode is terminated or truncated.
+        if other.is_terminated:
+            self.is_terminated = True
+        elif other.is_truncated:
+            self.is_truncated = True
+
+        # Copy over the current buffer values from the new episode chunk.
+        self._agent_buffered_actions = copy.deepcopy(other._agent_buffered_actions)
+        self._agent_buffered_rewards = other._agent_buffered_rewards.copy()
+        self._agent_buffered_extra_model_outputs = copy.deepcopy(
+            other._agent_buffered_extra_model_outputs
+        )
+
+        # Validate.
+        self.validate()
+
```

Comment:
AH, I have waited so long for this. Can't wait to try out the replay buffer now :)

Reply:
Yeah, hang in there. Almost there :)